File: 剑指offer/13. 机器人的运动范围.py
#######解法一：DFS(Deep Frist Search)

class Solution:
    def movingCount(self, m: int, n: int, k: int) -> int:
        def dfs(i, j, si, sj):
            if not 0 <= i < m or not 0 <= j < n or k < si + sj or (i, j) in visited: return 0
            visited.add((i,j))
            return 1 + dfs(i + 1, j, si + 1 if (i + 1) % 10 else si - 8, sj) + dfs(i, j + 1, si, sj + 1 if (j + 1) % 10 else sj - 8)

        visited = set()
        return dfs(0, 0, 0, 0)


#######解法一：BFS(Breadth Frist Search)

class Solution:
    def movingCount(self, m: int, n: int, k: int) -> int:
        queue, visited,  = [(0, 0, 0, 0)], set()
        while queue:
            i, j, si, sj = queue.pop(0)
            if not 0 <= i < m or not 0 <= j < n or k < si + sj or (i, j) in visited: continue
            visited.add((i,j))
            queue.append((i + 1, j, si + 1 if (i + 1) % 10 else si - 8, sj))
            queue.append((i, j + 1, si, sj + 1 if (j + 1) % 10 else sj - 8))
        return len(visited)


https://leetcode-cn.com/problems/ji-qi-ren-de-yun-dong-fan-wei-lcof/solution/mian-shi-ti-13-ji-qi-ren-de-yun-dong-fan-wei-dfs-b/

File: 剑指offer/C++_Version/13. 机器人的运动范围.cpp
// 方法1 DFS

class Solution {

public:
    int movingCount(int m, int n, int k)
    {
        vector<vector<bool>> visited(m, vector<bool>(n, 0));
        return dfs(m,n,k,visited,0,0,0,0);
    }


private:
    int dfs(int m, int n, int k, vector<vector<bool>> &visited, int i, int j, int si, int sj)
    {
        if(i<0 || i>=m || j<0 || j>=n || si + sj > k || visited[i][j])
            return false;
        visited[i][j] = true;  // c++中该处是这样处理的，没有用python set()那样的集合
        return 1 + dfs(m,n,k,visited,i+1,j, (i+1)%10!=0 ? si+1 : si-8,sj) +\
                   dfs(m,n,k,visited,i,j+1, si,(j+1)%10!=0 ? sj+1 : sj-8);
    }
};


// 方法2 BFS
class Solution {

public:
    int movingCount(int m, int n, int k)
    {  
        vector<vector<bool>> visited(m, vector<bool>(n, 0));
        queue<vector<int>> que;
        int res=0;
        que.push({0, 0, 0, 0});
        while(que.size()>0)
        {
            vector<int> x = que.front();
            que.pop();
            int i=x[0], j=x[1], si=x[2], sj=x[3];
            if(i<0 || i>=m || j<0 || j>=n || si+sj>k || visited[i][j])
                continue;
            visited[i][j] = true;  // c++中没有像python中那样, (i, j) in visited: 的写法。所以需要转个弯这样写
            res++;
            que.push({i+1, j, (i+1)%10 ? si+1 : si-8, sj});
            que.push({i, j+1, si, (j+1)%10 ? sj+1 : sj-8});
        }
        return res;
    }
};
